packages/realm/src/IterableWeakSet.ts

```diff
@@ -0,0 +1,128 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { binding } from "./internal";
+
+/**
+ * Implements an iterable `WeakSet`.
+ * @internal
+ */
+export class IterableWeakSet<T extends object> implements Set<T> {
+  private internal: Set<binding.WeakRef<T>>;
+  constructor(values?: readonly T[] | null) {
+    this.internal = new Set(values ? values.map((value) => new binding.WeakRef(value)) : undefined);
+  }
+
+  /**
+   * Deletes all weak references that no longer points to an object
+   */
+  clean() {
+    for (const ref of this.internal) {
+      if (!ref.deref()) {
+        this.internal.delete(ref);
```

Comment:
I also like that I can delete while iterating without having to keep track of the shifting offsets as I need to do with an array.

Reply:
The suggestion to use an Array only applied in the limited context of the other review where you never did partial removals. Doing partial removals with an Array is much more complicated, and something that makes sense with a Set. But in the other review your weren't actually benefiting from the "setness".